streamflow-api.yaml

```diff
@@ -0,0 +1,768 @@
+
+    # Lists (collections) of hydromodel outputs
+
+    HydromodelOutputListItem:
+      description: Representation of a single item in a list of hydromodel outputs.
+      type: object
+      allOf:
+        - type: object
+          properties:
+            uri:
+              type: string
+              description: URI of hydromodel output
+              example: 'http://apis.paccifcclimate.org/hydromodel_output/4'
+        - $ref: '#/components/schemas/HypermediaLinkList'
+        - $ref: '#/components/schemas/HydromodelOutputState'
+
+    HydromodelOutputList:
+      description: List of hydromodel outputs in collection.
+      type: array
+      items:
+        $ref: '#/components/schemas/HydromodelOutputListItem'
+      example:
+        - uri: 'http://apis.paccifcclimate.org/hydromodel_output/3'
+          model: "CanESM2"
+          experiment: "rcp45"
+        - uri: 'http://apis.paccifcclimate.org/hydromodel_output/4'
+          model: "CanESM2"
+          experiment: "rcp85"
+
+    ### Streamflow orders
+
+    # Individual streamflow orders
+
+    StreamflowOrderStateNew:
+      description: Defines a new order for a streamflow result set.
+      type: object
+      properties:
+        hydromodel_output_id:
+          $ref: '#/components/schemas/HydromodelOutputId'
+        longitude:
+          type: number
+          example: -123.5
+        latitude:
+          type: number
+          example: 48.5
+        notification:
+          type: object
+          properties:
+            method:
+              type: string
+              enum: [email]
+            address:
+              type: string
+              example: example@uvic.ca
+
+    StreamflowOrderState:
+      description: State for a streamflow order
+      allOf:
+        - type: object
+          properties:
+            id:
+              type: integer
+              example: 5
+            status:
+              type: string
+              enum: [pending, cancelled, fulfilled, deprecated]
+              example: fulfilled
+        - $ref: '#/components/schemas/StreamflowOrderStateNew'
+
+    StreamflowOrderHypermedia:
+      description: Hypermedia for a streamflow order
+      type: object
+      properties:
+        links:
+          allOf:
+            - $ref: '#/components/schemas/HypermediaLinkList'
+            - example:
+                - rel: 'self'
+                  uri: 'http://apis.paccifcclimate.org/streamflow/orders/5'
+                - rel: 'cancel'
+                  uri: 'http://apis.paccifcclimate.org/streamflow/orders/5'
+                - rel: 'hydromodel_output'
+                  uri: 'http://apis.paccifcclimate.org/hydromodel_outputs/4'
+                - rel: 'results/metadata'
+                  uri: 'http://apis.paccifcclimate.org/streamflow/results/6'
+                - rel: 'results/annual-max'
+                  uri: 'http://apis.paccifcclimate.org/streamflow/results/6/annual-max'
```

Comment:
So it's the RESTy thing to do to include these links to annual cycle, annual mean, and annual max even though they don't exist when the order is created and will be 404s?

Reply:
Great observation. REST has nothing to say about this either way. A link to an absent resource isn't really non-RESTful, but it's not really very logical either. A good rule is don't have links that you can know in advance are invalid.